Move design to MVVM

mnadareski opened the pull request:
This PR has the overall goal of moving to an MVVM pattern (fixes #22).

As a part of this, the following changes are implemented:
- Removed all references to Settings object in anything but UI code
- Removed all knowledge of the UI from the view-model and model code
- Added subscriber events for model and view-model code
- Removed UI timer for forcing updates
- Provided more methods to safely set values going down the chain
- Separated out the GUI and common code into separate libraries for future use

Notably, on my development machine, this decreased CPU usage by a fair bit, especially at idle.

As a bonus, since it came up during the refactoring and testing, hidden tracks can now play properly and are controlled by an internal flag set by the settings with the current UI (fixes #10). This ended up being a pretty major thing to address as part of this PR, because a lot of the behavior that was causing hidden tracks to render incorrectly was tied to a poor implementation of track loading and switching.

mnadareski committed:
Make SoundOutput have events

mnadareski committed:
Make OpticalDisc have events

mnadareski committed:
Remove need for UI timer for updates

mnadareski committed:
Move all image logic to factory

mnadareski committed:
Make view-model safer

mnadareski committed:
Move methods to better locations

mnadareski committed:
Consolidate view init code

mnadareski committed:
View-model and model shouldn't know about settings

mnadareski committed:
Fix minor playback issues

mnadareski commented:
Please note that I am opening this in a draft state because I am not 100% confident that I applied the MVVM pattern to the code correctly.

mnadareski committed:
Remove needless method calls

mnadareski committed:
Include SoundOutput in force refresh

mnadareski committed:
Split common and GUI code into separate projects

mnadareski committed:
Fix volume setting; move view model to common

mnadareski committed:
Ensure access modifiers set properly

mnadareski committed:
Fix last access modifier

mnadareski commented:
Marking this as ready to review finally, because I feel like I've isolated the pieces properly at this point. All of my local tests are working as expected, with no regressions that I have been able to spot.

mnadareski committed:
Fix rendering of hidden tracks

mnadareski committed:
Safer track switching

mnadareski committed:
Fix off-by-one display total tracks if hidden track exists

mnadareski committed:
Fix improper use of field in PlayerViewModel

mnadareski committed:
Persist view model between theme swaps

mnadareski committed:
Fix SetTotalIndexes for corner case

mnadareski committed:
Handle data track corner cases

mnadareski committed:
Make de-emphasis a separate method

mnadareski committed:
Extract out another helper method

mnadareski commented:
All future work items have been added as appropriately-tagged issues. Nothing else will be added outside of requested changes.

claunia commented on RedBookPlayer.Common/Discs/CompactDisc.cs:
```diff
@@ -264,31 +418,63 @@ public override bool PreviousIndex(bool changeTrack)
+        /// Set the value for loading data tracks
+        /// </summary>
+        /// <param name="load">True to enable loading data tracks, false otherwise</param>
+        public void SetLoadDataTracks(bool load) => _loadDataTracks = load;
```
Use properties.

claunia commented on RedBookPlayer.Common/Discs/CompactDisc.cs:
```diff
@@ -264,31 +418,63 @@ public override bool PreviousIndex(bool changeTrack)
+        /// Set the value for loading hidden tracks
+        /// </summary>
+        /// <param name="load">True to enable loading hidden tracks, false otherwise</param>
+        public void SetLoadHiddenTracks(bool load) => _loadHiddenTracks = load;
```
Use properties.

claunia commented on RedBookPlayer.Common/Discs/CompactDisc.cs:
```diff
@@ -6,11 +6,12 @@
 using static Aaru.Decoders.CD.FullTOC;
 
-namespace RedBookPlayer.Discs
+namespace RedBookPlayer.Common.Discs
```
Is this a view, a model, or a viewmodel?

claunia commented on RedBookPlayer.Common/Discs/OpticalDisc.cs:
```diff
@@ -1,9 +1,10 @@
+namespace RedBookPlayer.Common.Discs
 {
-    public abstract class OpticalDisc
+    public abstract class OpticalDisc : ReactiveObject
```
Shouldn't it be called OpticalDiscBase?

claunia commented on RedBookPlayer.Common/Discs/OpticalDisc.cs:
```diff
@@ -1,9 +1,10 @@
+using ReactiveUI;
 
-namespace RedBookPlayer.Discs
+namespace RedBookPlayer.Common.Discs
```
View, Model or ViewModel?

claunia commented on RedBookPlayer.Common/Discs/OpticalDiscFactory.cs:
```diff
@@ -1,17 +1,57 @@
+using Aaru.Filters;
 
-namespace RedBookPlayer.Discs
+namespace RedBookPlayer.Common.Discs
```
Factories should go in factories namespace and folder imo.

claunia commented on RedBookPlayer.Common/PlayerViewModel.cs:
```diff
@@ -0,0 +1,328 @@
+using ReactiveUI;
+using RedBookPlayer.Common.Hardware;
+
+namespace RedBookPlayer.Common
```
Namespace, namespace

claunia commented on RedBookPlayer.GUI/MainWindow.xaml.cs:
```diff
@@ -5,6 +5,7 @@
 using Avalonia.Markup.Xaml;
+using RedBookPlayer.Common;
 
 namespace RedBookPlayer.GUI
```
namespace namespace. Views views.

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
         /// Load an image from the path
         /// </summary>
         /// <param name="path">Path to the image to load</param>
         public async Task<bool> LoadImage(string path)
```
This should go in the viewmodel.

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
+        /// <returns>String representing the digits for the frontend</returns>
+        private string GenerateDigitString()
+        {
+            // If the disc isn't initialized, return all '-' characters
```
This should go in the viewmodel.

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
+        /// Generate a path selection dialog box
+        /// </summary>
+        /// <returns>User-selected path, if possible</returns>
+        private async Task<string> GetPath()
```
-> Viewmodel

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
-        /// <param name="xaml">XAML data representing the theme, null for default</param>
-        private void InitializeComponent(string xaml)
+        /// <returns>ulong representing the current sector time</returns>
+        private ulong GetCurrentSectorTime()
```
-> Viewmodel

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -181,20 +219,38 @@ private void InitializeDigits()
+        /// Update the UI from the view-model
         /// </summary>
-        private void UpdateView(object sender, ElapsedEventArgs e)
+        private void PlayerViewModelStateChanged(object sender, PropertyChangedEventArgs e)
```
-> Viewmodel

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
+        }
 
-        public void StopButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = null;
+        public void StopButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Stop();
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
+        public void PauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Pause();
 
-        public void PlayPauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = !(PlayerViewModel.Playing ?? false);
+        public void PlayPauseButton_Click(object sender, RoutedEventArgs e)
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
+        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Play();
 
-        public void PauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = false;
+        public void PauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Pause();
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
         }
 
-        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = true;
+        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Play();
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
         public void RewindButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Rewind();
 
-        public void VolumeUpButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Volume++;
+        public void VolumeUpButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetVolume(PlayerViewModel.Volume + 1);
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
 
         public void RewindButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Rewind();
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
+        public void VolumeUpButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetVolume(PlayerViewModel.Volume + 1);
 
-        public void VolumeDownButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Volume--;
+        public void VolumeDownButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetVolume(PlayerViewModel.Volume - 1);
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
         public void MuteToggleButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ToggleMute();
 
-        public void EnableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ApplyDeEmphasis = true;
+        public void EnableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(true);
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
+        public void EnableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(true);
 
-        public void DisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ApplyDeEmphasis = false;
+        public void DisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(false);
```
Binding

claunia commented on RedBookPlayer.GUI/PlayerView.xaml.cs:
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
+        public void DisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(false);
 
-        public void EnableDisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ApplyDeEmphasis = !PlayerViewModel.ApplyDeEmphasis;
+        public void EnableDisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(!PlayerViewModel.ApplyDeEmphasis);
```
Binding

mnadareski committed:
Move more logic down the chain

mnadareski committed:
Factories and base class renames

mnadareski committed:
Use properties in CompactDisc

mnadareski committed:
Move image loading to view model

mnadareski committed:
`Common` -> `Models`

mnadareski committed:
Add Views namespace

mnadareski committed:
Fix build because of missing namespaces

mnadareski committed:
Fix UI thread issues

mnadareski committed:
Update gitmodules

The standard version of the library does NOT support OpenAL on both Windows and Linux. This branch supports it and has been open as a PR on the main netstandard support branch for a while now.
